File: web/src/protocol.ts
export type StoneColor = 'black' | 'white';
export type GameOutcome = 'black_won' | 'white_won' | 'draw';

export interface CheckpointInfo {
  architecture: string;
  epoch: number;
  model_digest: string;
}

export interface Cell {
  row: number;
  column: number;
}

export interface Stone extends Cell {
  color: StoneColor;
}

export interface MoveStats extends Cell {
  prior: number;
  visits: number;
  mean_value: number | null;
}

export type ClientMessage =
  | { type: 'new_game'; human_color: StoneColor }
  | { type: 'restore_game'; human_color: StoneColor; moves: Cell[] }
  | { type: 'start_search'; position_id: number; simulations: number }
  | { type: 'stop_search'; position_id: number }
  | { type: 'play'; position_id: number; row: number; column: number }
  | { type: 'choose_network_move'; position_id: number; temperature: number };

export interface HelloMessage {
  type: 'hello';
  protocol_version: number;
  board_size: number;
  compute_device: string;
  checkpoint: CheckpointInfo;
  max_search_simulations: number;
  default_search_simulations: number;
  c_puct: number;
  snapshot_interval_ms: number;
}

export interface PositionMessage {
  type: 'position';
  position_id: number;
  ply: number;
  human_color: StoneColor;
  to_move: StoneColor;
  stones: Stone[];
  last_move: Cell | null;
  outcome: GameOutcome | null;
  carried_visits: number;
}

export interface SearchStatusMessage {
  type: 'search_status';
  position_id: number;
  analysis_id: number;
  searched_simulations: number;
  target_simulations: number;
  running: boolean;
}

export interface SearchSnapshotMessage {
  type: 'search_snapshot';
  position_id: number;
  analysis_id: number;
  searched_simulations: number;
  carried_visits: number;
  total_visits: number;
  target_simulations: number;
  elapsed_ms: number;
  simulations_per_second: number;
  network_value: number;
  search_value: number | null;
  moves: MoveStats[];
  complete: boolean;
}

export interface ErrorMessage {
  type: 'error';
  code: string;
  message: string;
  recoverable: boolean;
}

export type ServerMessage =
  | HelloMessage
  | PositionMessage
  | SearchStatusMessage
  | SearchSnapshotMessage
  | ErrorMessage;

export const PROTOCOL_VERSION = 2;
export const BOARD_SIZE = 19;
export const COLUMNS = 'ABCDEFGHJKLMNOPQRST';

export function restoreGameCommand(
  position: PositionMessage | null,
  fallbackHumanColor: StoneColor,
): Extract<ClientMessage, { type: 'restore_game' }> {
  return {
    type: 'restore_game',
    human_color: position?.human_color ?? fallbackHumanColor,
    moves: position?.stones.map(({ row, column }) => ({ row, column })) ?? [],
  };
}

export function cellKey(cell: Cell): string {
  return `${cell.row}:${cell.column}`;
}

export function moveName(cell: Cell): string {
  return `${COLUMNS[cell.column] ?? '?'}${BOARD_SIZE - cell.row}`;
}

export function visitFraction(move: MoveStats, snapshot: SearchSnapshotMessage): number {
  return snapshot.total_visits > 0 ? move.visits / snapshot.total_visits : 0;
}

/** Mirrors alz::engine::apply_temperature, including its last-wins tie break. */
export function temperatureProbabilities(
  moves: MoveStats[],
  temperature: number,
): Map<string, number> {
  const result = new Map<string, number>();
  if (moves.length === 0) return result;

  let maxIndex = 0;
  for (let index = 1; index < moves.length; index += 1) {
    if ((moves[index]?.visits ?? 0) >= (moves[maxIndex]?.visits ?? 0)) maxIndex = index;
  }
  const maxVisits = moves[maxIndex]?.visits ?? 0;
  if (maxVisits === 0) {
    for (const move of moves) result.set(cellKey(move), 0);
    return result;
  }
  if (temperature === 0) {
    moves.forEach((move, index) => result.set(cellKey(move), index === maxIndex ? 1 : 0));
    return result;
  }

  const weights = moves.map((move) =>
    move.visits === 0 ? 0 : Math.exp(Math.log(move.visits / maxVisits) / temperature),
  );
  const sum = weights.reduce((total, weight) => total + weight, 0);
  moves.forEach((move, index) => result.set(cellKey(move), (weights[index] ?? 0) / sum));
  return result;
}

File: crates/demo-server/src/protocol.rs
use serde::{Deserialize, Serialize};

pub const PROTOCOL_VERSION: u32 = 2;

#[derive(Clone, Copy, Debug, Deserialize, Eq, PartialEq, Serialize)]
#[serde(rename_all = "snake_case")]
pub enum StoneColor {
    Black,
    White,
}

#[derive(Clone, Copy, Debug, Eq, PartialEq, Serialize)]
#[serde(rename_all = "snake_case")]
pub enum GameOutcome {
    BlackWon,
    WhiteWon,
    Draw,
}

#[derive(Clone, Debug, Deserialize, PartialEq)]
#[serde(tag = "type", rename_all = "snake_case", deny_unknown_fields)]
pub enum ClientMessage {
    NewGame {
        human_color: StoneColor,
    },
    RestoreGame {
        human_color: StoneColor,
        moves: Vec<Cell>,
    },
    StartSearch {
        position_id: u64,
        simulations: u32,
    },
    StopSearch {
        position_id: u64,
    },
    Play {
        position_id: u64,
        row: u8,
        column: u8,
    },
    ChooseNetworkMove {
        position_id: u64,
        temperature: f32,
    },
}

#[derive(Clone, Debug, PartialEq, Serialize)]
pub struct CheckpointInfo {
    pub architecture: String,
    pub epoch: usize,
    pub model_digest: String,
}

#[derive(Clone, Copy, Debug, Deserialize, PartialEq, Serialize)]
pub struct Cell {
    pub row: u8,
    pub column: u8,
}

#[derive(Clone, Copy, Debug, PartialEq, Serialize)]
pub struct Stone {
    pub row: u8,
    pub column: u8,
    pub color: StoneColor,
}

#[derive(Clone, Debug, PartialEq, Serialize)]
pub struct MoveStats {
    pub row: u8,
    pub column: u8,
    pub prior: f32,
    pub visits: u32,
    pub mean_value: Option<f32>,
}

#[derive(Clone, Debug, PartialEq, Serialize)]
#[serde(tag = "type", rename_all = "snake_case")]
pub enum ServerMessage {
    Hello {
        protocol_version: u32,
        board_size: usize,
        compute_device: String,
        checkpoint: CheckpointInfo,
        max_search_simulations: u32,
        default_search_simulations: u32,
        c_puct: f32,
        snapshot_interval_ms: u64,
    },
    Position {
        position_id: u64,
        ply: usize,
        human_color: StoneColor,
        to_move: StoneColor,
        stones: Vec<Stone>,
        last_move: Option<Cell>,
        outcome: Option<GameOutcome>,
        carried_visits: u32,
    },
    SearchStatus {
        position_id: u64,
        analysis_id: u64,
        searched_simulations: u32,
        target_simulations: u32,
        running: bool,
    },
    SearchSnapshot {
        position_id: u64,
        analysis_id: u64,
        searched_simulations: u32,
        carried_visits: u32,
        total_visits: u32,
        target_simulations: u32,
        elapsed_ms: u64,
        simulations_per_second: f64,
        network_value: f32,
        search_value: Option<f32>,
        moves: Vec<MoveStats>,
        complete: bool,
    },
    Error {
        code: &'static str,
        message: String,
        recoverable: bool,
    },
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn protocol_uses_stable_discriminators_and_field_names() {
        let command = serde_json::from_str::<ClientMessage>(
            r#"{"type":"start_search","position_id":7,"simulations":2000}"#,
        )
        .unwrap();
        assert_eq!(
            command,
            ClientMessage::StartSearch {
                position_id: 7,
                simulations: 2_000,
            }
        );

        let restore = serde_json::from_str::<ClientMessage>(
            r#"{"type":"restore_game","human_color":"white","moves":[{"row":9,"column":9},{"row":9,"column":10}]}"#,
        )
        .unwrap();
        assert_eq!(
            restore,
            ClientMessage::RestoreGame {
                human_color: StoneColor::White,
                moves: vec![Cell { row: 9, column: 9 }, Cell { row: 9, column: 10 },],
            }
        );

        let message = ServerMessage::SearchStatus {
            position_id: 7,
            analysis_id: 3,
            searched_simulations: 512,
            target_simulations: 2_000,
            running: true,
        };
        assert_eq!(
            serde_json::to_value(message).unwrap(),
            serde_json::json!({
                "type": "search_status",
                "position_id": 7,
                "analysis_id": 3,
                "searched_simulations": 512,
                "target_simulations": 2_000,
                "running": true,
            })
        );
    }

    #[test]
    fn hello_reports_the_actual_compute_device() {
        let message = ServerMessage::Hello {
            protocol_version: PROTOCOL_VERSION,
            board_size: 19,
            compute_device: "CPU".to_owned(),
            checkpoint: CheckpointInfo {
                architecture: "kata_v1".to_owned(),
                epoch: 38,
                model_digest: "abcdef012345".to_owned(),
            },
            max_search_simulations: 10_000,
            default_search_simulations: 2_000,
            c_puct: 1.0,
            snapshot_interval_ms: 100,
        };
        let json = serde_json::to_value(message).unwrap();

        assert_eq!(json["compute_device"], "CPU");
    }
}
